drivers/flash/nrf_qspi_nor.c

```diff
@@ -503,27 +505,86 @@ static inline int qspi_nor_read_id(struct device *dev,
 	return 0;
 }
 
+static inline nrfx_err_t read_non_aligned(struct device *dev, off_t addr,
+					  void *dest, size_t size)
+{
+	uint8_t __aligned(4) buf[WORD_SIZE * 2];
+	uint8_t *dptr = dest;
+
+	off_t flash_prefix = WORD_SIZE - (addr % WORD_SIZE);
+
+	if (flash_prefix > size) {
+		flash_prefix = size;
+	}
+
+	off_t dest_prefix = (WORD_SIZE - (off_t)dptr % WORD_SIZE) % WORD_SIZE;
+
+	if (dest_prefix > size) {
+		dest_prefix = size;
+	}
+
+	off_t flash_suffix = (size - flash_prefix) % WORD_SIZE;
+	off_t flash_middle = size - flash_prefix - flash_suffix;
+	off_t dest_middle = size - dest_prefix -
+			    (size - dest_prefix) % WORD_SIZE;
+
+	if (flash_middle > dest_middle) {
+		flash_middle = dest_middle;
+		flash_suffix = size - flash_prefix - flash_middle;
+	}
+
+	nrfx_err_t res = NRFX_SUCCESS;
+
+	/* read from aligned flash to aligned memory */
+	if (flash_middle != 0) {
+		res = nrfx_qspi_read(dptr + dest_prefix, flash_middle,
+				     addr + flash_prefix);
+		qspi_wait_for_completion(dev, res);
+		if (res != NRFX_SUCCESS) {
+			return res;
+		}
+	}
+
+	/* perform shift in RAM */
+	if (flash_prefix - dest_prefix != 0) {
```

Comment:
The arithmetic expression should be in parentheses.

Reply:
Added.